zookeeper-server/src/main/java/org/apache/zookeeper/server/util/RequestPathMetricsCollector.java

```diff
@@ -0,0 +1,385 @@
+        switch (requestType) {
+            case ZooDefs.OpCode.sync:
+            case ZooDefs.OpCode.create:
+            case ZooDefs.OpCode.create2:
+            case ZooDefs.OpCode.createContainer:
+            case ZooDefs.OpCode.delete:
+            case ZooDefs.OpCode.deleteContainer:
+            case ZooDefs.OpCode.setData:
+            case ZooDefs.OpCode.reconfig:
+            case ZooDefs.OpCode.setACL:
+            case ZooDefs.OpCode.multi:
+            case ZooDefs.OpCode.check:
+                return true;
+        }
+        return false;
+    }
+
+    static String trimPathDepth(String path, int maxDepth) {
+        int count = 0;
+        String pathSeparator = "/";
+        StringBuilder sb = new StringBuilder();
+        StringTokenizer pathTokenizer = new StringTokenizer(path, pathSeparator);
+        while (pathTokenizer.hasMoreElements() && count++ < maxDepth) {
+            sb.append(pathSeparator);
+            sb.append(pathTokenizer.nextToken());
+        }
+        path = sb.toString();
+        return path;
+    }
+
+    @Override
+    protected void finalize() {
+        shutdown();
+    }
+        
+    public void shutdown(){
+        if (disabled) return;
+
+        LOG.info("shutdown scheduledExecutor");
+        scheduledExecutor.shutdownNow();
+    }
+    
+    public void start() {
+        if (disabled) return;
+
+        LOG.info("Start the RequestPath collector");
+        immutableRequestsMap.forEach((opType, pathStatsQueue) -> pathStatsQueue.start());
+        
+        // Schedule to log the top used read/write paths every 5 mins
+        scheduledExecutor.scheduleWithFixedDelay(() -> {
+            LOG.info("%nHere are the top Read paths:");
+            logTopPaths(aggregatePaths(4, queue -> !queue.isWriteOperation()),
+                entry -> LOG.info(entry.getKey() + " : " + entry.getValue()));
+            LOG.info("%nHere are the top Write paths:");
+            logTopPaths(aggregatePaths(4, queue -> queue.isWriteOperation()),
+                entry -> LOG.info(entry.getKey() + " : " + entry.getValue()));
+        }, COLLECTOR_INITIAL_DELAY, COLLECTOR_DELAY, TimeUnit.MINUTES);
+    }
+
+    /**
+     * The public interface of the buffer. FinalRequestHandler will call into this for
+     * each request that has a path and this needs to be fast. we sample the path so that
+     * we don't have to store too many paths in memory
+     */
+    public void registerRequest(int type, String path) {
+        if (sampler.nextFloat() <= REQUEST_PREPROCESS_SAMPLE_RATE) {
+            PathStatsQueue pathStatsQueue = immutableRequestsMap.get(Request.op2String(type));
+            if (pathStatsQueue != null) {
+                pathStatsQueue.registerRequest(path);
+            } else {
+                LOG.error("We should not handle {}", type);
+            }
+        }
+    }
+
+    public void dumpTopRequestPath(PrintWriter pwriter, String requestTypeName, int queryMaxDepth) {
```

Comment:
where is this method used ? in an admin command ?

Reply:
That's a yes! It's used in a new "path" command we added internally. Once this patch landed, I'll that command. I tried to scope this patch to path stats collection only to avoid it bloated and make it easier to review.